generator/tree-gen-cpp.cpp

```diff
@@ -1072,6 +1081,172 @@ void generate_dumper_class(
+                            R"(        out << "\"{0}\":\"{1}\"";)""\n",
+                            attrib.name,
+                            (attrib.ext_type == One || attrib.ext_type == Link) ? "!MISSING" : "-");
+                        fmt::print(source,
+                            "    }} else {{\n");
+                        if (attrib.ext_type == Link || attrib.ext_type == OptLink) {
+                            fmt::print(source,
+                                "        if (!in_link) {{\n"
+                                "            in_link = true;\n");
+                            if (attrib.type == Prim) {
+                                fmt::print(source,
+                                    R"(            out << "\"{0}\":";)""\n"
+                                    R"(            if (!node.{0}.empty()) {{)""\n"
+                                    R"(                node.{0}->dump_json(out);)""\n"
+                                    R"(            }})""\n",
+                                    attrib.name);
+                            } else {
+                                fmt::print(source,
+                                    R"(            out << "\"{0}\":";)""\n"
+                                    R"(            node.{0}.visit(*this);)""\n",
+                                attrib.name);
+                            }
+                            fmt::print(source,
+                                R"(            in_link = false;)""\n"
+                                R"(        }} else {{)""\n"
+                                R"(            out << "\"{0}\":\"...\"";)""\n"
+                                R"(        }})""\n",
+                                attrib.name);
+                        } else if (attrib.type == Prim) {
+                            fmt::print(source,
+                                R"(      out << "\"{0}\":";)""\n"
+                                R"(      if (!node.{0}.empty()) {{)""\n"
+                                R"(          node.{0}->dump_json(out);)""\n"
+                                R"(      }})""\n",
+                                attrib.name);
+                        } else {
+                            fmt::print(source,
+                                R"(      out << "\"{0}\":";)""\n"
+                                R"(      node.{0}.visit(*this);)""\n",
+                                attrib.name);
+                        }
+                        fmt::print(source,
+                            "    }}\n");
+                        break;
+
+                    case Any:
+                    case Many:
+                        fmt::print(source,
+                            "    if (node.{0}.empty()) {{\n"
+                            R"(        out << "\"{0}\":\"{1}\"";)""\n"
+                            "    }} else {{\n"
+                            R"(        out << "\"{0}\":[";)""\n"
+                            "        bool first_element = true;\n"
```

Comment:
Jesus Christ! 😮 
I find this ugly and difficult to read. Not you?
I really wish we (Jeroen) had chosen a better way to generate C++ than use C++ itself...
Like, for instance, a scripting library made for code generation, or even better a template language.

Meanwhile, is there here a way to avoid
- having to repeat curly braces?
- having each line start with a quote?
- having the \n everywhere?
- having every quote escaped with a backslash?

Reply:
Hi Pablo,

Thanks for the comment. There is something in that piece of code that I was trying to avoid, mixing literals and raw literals `R"()"`, but I forgot doing it there.

So, if you see, for my `generate` function, I tried using `fmt::print` hoping the generated code could be a bit more legible, but I'm afraid I haven't managed to do it. If you notice, the rest of the code uses `iostream`s.

Now, answering your questions:
- Repeating curly braces is a `fmt` thing, because braces are used as placeholders (i.e., `{}` points to a parameter), so if you want to print a brace, you have to double it.
- There would be a way of avoiding each line to start with a quote, and using the `\n` everywhere, but that would imply to write the pieces of code unindented with respect to the real code. For example:

```
                case Many:
                    fmt::print(source, R"(
    if (node.{0}.empty()) {{
        out << "\"{0}\":\"{1}\"";
    }} else {{
        out << "\"{0}\":[";
        bool first_element = true;)"
```

- If you see the code above, because we are using raw literals `R"()"`, we don't need to escape characters (like `out << "`). What you write within the raw literal is what is written out to a C++ file. However, that C++ code can contain escaped double quotes (like `out << "\"`).

Reply:
Regarding my last comment, after rewriting some code using a single string for each block, I saw that the reading was really awful, so I've decided to keep it as it is.